File: app/src/main/java/ru/art2000/calculator/currency_converter/view/CurrencyConverterFragment.java
package ru.art2000.calculator.currency_converter.view;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffColorFilter;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.view.menu.ActionMenuItemView;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import ru.art2000.calculator.MainActivity;
import ru.art2000.calculator.R;
import ru.art2000.calculator.currency_converter.model.LoadingState;
import ru.art2000.calculator.currency_converter.view_model.CurrencyConverterModel;
import ru.art2000.extensions.IReplaceable;
import ru.art2000.extensions.ReplaceableFragment;
import ru.art2000.helpers.AndroidHelper;
import ru.art2000.helpers.CurrencyValuesHelper;

public class CurrencyConverterFragment extends ReplaceableFragment {

    public Context mContext;
    public CurrencyListAdapter adapter = null;
    private TextView emptyView;
    private RecyclerView recycler;
    private MainActivity parent;
    private View v = null;

    private SwipeRefreshLayout refresher;
    private boolean isUpdating;
    private boolean didFirstUpdate;
    private Toolbar mToolbar;
    private String titleUpdatedString;


    private CurrencyConverterModel model;


    private void setRefreshStatus(boolean status) {
        if (refresher != null) {
            refresher.setRefreshing(status);
            isUpdating = status;
        }
    }

    public void scrollToTop() {
        if (recycler != null)
            recycler.smoothScrollToPosition(0);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == 666) {
            if (resultCode == 1) {
//                adapter.getDataFromDB();
//                toggleEmptyView();
            }
        }
    }

    @SuppressLint("InflateParams")
    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        if (v == null) {
            mContext = getActivity();
            parent = (MainActivity) requireActivity();
            model = new CurrencyConverterModel(parent.getApplication());


            titleUpdatedString = mContext.getString(R.string.updated);

            v = inflater.inflate(R.layout.currency_layout, null);
            recycler = v.findViewById(R.id.currency_list);
            emptyView = v.findViewById(R.id.empty_tv);
            mToolbar = v.findViewById(R.id.toolbar);
            mToolbar.inflateMenu(R.menu.currencies_converter_menu);

            LinearLayoutManager llm = new LinearLayoutManager(mContext);
            recycler.setLayoutManager(llm);
            adapter = new CurrencyListAdapter(mContext);
            recycler.setAdapter(adapter);

            recycler.setOnFocusChangeListener((v, hasFocus) ->
                    adapter.removeEditText());

            int colorAccent = AndroidHelper.getColorAttribute(mContext, R.attr.colorAccent);

            refresher = v.findViewById(R.id.refresher);
            refresher.setColorSchemeColors(colorAccent);
            refresher.setProgressViewEndTarget(true, refresher.getProgressViewEndOffset());
            refresher.setOnRefreshListener(this::updateData);

            ActionMenuItemView editMenuItem = v.findViewById(R.id.edit_currencies);
            editMenuItem.getItemData().getIcon().setColorFilter(
                    new PorterDuffColorFilter(colorAccent, PorterDuff.Mode.SRC_ATOP));
            editMenuItem.setOnClickListener(v -> {
                adapter.removeEditText();
                Intent intent = new Intent(getActivity(), CurrenciesSettingsActivity.class);
                startActivityForResult(intent, 666);
            });
        }

        return v;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        model.getLoadingState().observe(getViewLifecycleOwner(), loadingState -> {
            setRefreshStatus(loadingState == LoadingState.LOADING_STARTED);

            if (loadingState == LoadingState.UNINITIALISED || loadingState == LoadingState.LOADING_ENDED)
                return;

            int messageId;
            switch (loadingState) {
                case LOADING_STARTED:
                    messageId = R.string.currencies_update_toast;
                    break;
                case NETWORK_ERROR:
                    messageId = R.string.currencies_no_internet;
                    break;
                default:
                    messageId = R.string.currencies_update_failed;
                    break;
            }

            Toast.makeText(mContext, messageId, Toast.LENGTH_SHORT).show();
        });


        model.getUpdateDate().observe(getViewLifecycleOwner(), this::setCurrenciesUpdateDate);
        model.getVisibleList().observe(getViewLifecycleOwner(), currencyItems -> {

            adapter.setNewData(currencyItems);

            if (currencyItems.isEmpty()) {
                emptyView.setVisibility(View.VISIBLE);
                refresher.setVisibility(View.GONE);
            } else {
                emptyView.setVisibility(View.GONE);
                refresher.setVisibility(View.VISIBLE);
            }
        });
    }

    private void setCurrenciesUpdateDate(String date) {
        mToolbar.setTitle(titleUpdatedString + " " + date);
    }

    @Override
    public void onPause() {
        adapter.removeEditText();
        super.onPause();
    }

    @Override
    public void onResume() {
        super.onResume();
        setRefreshStatus(isUpdating);
    }

    private void updateData() {
        didFirstUpdate = true;
        new Thread(() -> model.loadData()).start();
    }

    @Override
    protected void onShown(@Nullable IReplaceable previousReplaceable) {
        if (!didFirstUpdate) {
            updateData();
        }

        parent.changeStatusBarColor(false);
    }

    @Override
    public int getOrder() {
        return 0;
    }

    @Override
    public int getReplaceableId() {
        return R.id.navigation_currency;
    }

    @Override
    public int getIcon() {
        return R.drawable.ic_currency;
    }

    @Override
    public int getTitle() {
        return R.string.title_currency;
    }

}

File: currency/src/androidMain/kotlin/ru/art2000/calculator/currency/view/CurrencyConverterFragment.kt
package ru.art2000.calculator.currency.view

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.appcompat.view.menu.ActionMenuItemView
import androidx.fragment.app.viewModels
import com.google.android.material.datepicker.*
import com.google.android.material.snackbar.Snackbar
import dagger.hilt.android.AndroidEntryPoint
import dev.androidbroadcast.vbpd.viewBinding
import ru.art2000.calculator.common.view.MainScreenFragment
import ru.art2000.calculator.common.view.createThemedSnackbar
import ru.art2000.calculator.currency.R
import ru.art2000.calculator.currency.databinding.CurrencyLayoutBinding
import ru.art2000.calculator.currency.model.LoadingState
import ru.art2000.calculator.currency.vm.CurrencyConverterModel
import ru.art2000.extensions.LocalDate
import ru.art2000.extensions.arch.launchRepeatOnStarted
import ru.art2000.extensions.getColorFromAttribute
import ru.art2000.extensions.kt.launchAndCollect
import ru.art2000.extensions.timeInMillis
import ru.art2000.extensions.views.*
import ru.art2000.calculator.common.R as CommonR

@AndroidEntryPoint
internal class CurrencyConverterFragment : MainScreenFragment(R.layout.currency_layout) {

    private val model by viewModels<CurrencyConverterModel>()
    private val binding by viewBinding(CurrencyLayoutBinding::bind)
    private var currenciesAdapter: CurrencyListAdapter? = null
    private var keyboardListenerSubscription: ListenerSubscription<Boolean>? = null
    private var updateSnackbar: Snackbar? = null

    override val topViews: List<View>
        get() = listOf(binding.root)

    override val bottomViews: List<View>
        get() = if (requireContext().isLandscape) listOf(binding.currencyList) else emptyList()

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.toolbar.inflateMenu(R.menu.currencies_converter_menu)
        currenciesAdapter = CurrencyListAdapter(requireContext(), model)

        binding.currencyList.apply {
            layoutManager = OrientationManger(requireContext()) {
                currenciesAdapter?.itemCount.let { it == null || it == 0 }
            }
            adapter = currenciesAdapter
            emptyViewGenerator = { _, _, _ ->
                createTextEmptyView(
                    requireContext(), R.string.empty_text_no_currencies_added
                )
            }
            addOrientationItemDecoration()
        }

        binding.refresher.apply {
            val colorAccent =
                requireContext().getColorFromAttribute(com.google.android.material.R.attr.colorSecondary)
            val circleBackground =
                requireContext().getColorFromAttribute(CommonR.attr.floatingViewBackground)
            setProgressBackgroundColorSchemeColor(circleBackground)
            setColorSchemeColors(colorAccent)
            setProgressViewEndTarget(true, progressViewEndOffset)
            setOnRefreshListener { model.loadData() }
        }

        val editMenuItem =
            binding.root.findViewById<ActionMenuItemView>(R.id.edit_currencies)
        editMenuItem.setOnClickListener {
            val intent = Intent(activity, CurrenciesSettingsActivity::class.java)
            startActivity(intent)
        }
        val selectDateMenuItem =
            binding.root.findViewById<ActionMenuItemView>(R.id.select_date)
        selectDateMenuItem.setOnClickListener {
            model.loadTimeIntervals { minDate, maxDate ->

                val minDateMillis = minDate.timeInMillis
                val maxDateMillis = maxDate.timeInMillis

                val minValidator = DateValidatorPointForward.from(minDateMillis)
                val maxValidator = DateValidatorPointBackward.before(maxDateMillis)

                val picker = MaterialDatePicker.Builder
                    .datePicker()
                    .setCalendarConstraints(
                        CalendarConstraints.Builder()
                            .setStart(minDateMillis)
                            .setEnd(maxDateMillis)
                            .setValidator(
                                CompositeDateValidator.allOf(
                                    listOf(minValidator, maxValidator)
                                )
                            ).build()
                    )
                    .build()
                picker.addOnPositiveButtonClickListener { selection ->
                    model.loadData(LocalDate(selection))
                }
                picker.show(requireActivity().supportFragmentManager, "CurrencyDatePicker")
            }
        }
        selectDateMenuItem.setOnLongClickListener {
            model.loadData()
            true
        }

        model.listenDateUpdate()

        launchRepeatOnStarted {
            launchAndCollect(model.loadingState) { applyLoadingState(it) }
            launchAndCollect(model.updateDate) { setCurrenciesUpdateDate(it) }
            launchAndCollect(model.visibleList) { currenciesAdapter?.setNewData(it) }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        model.stopListeningDateUpdate()
        currenciesAdapter = null
    }

    override fun onResume() {
        super.onResume()
        keyboardListenerSubscription =
            binding.root.addImeVisibilityListener { isVisible: Boolean ->
                if (!isVisible) {
                    currenciesAdapter?.removeEditText()
                }
            }
    }

    override fun onPause() {
        super.onPause()
        keyboardListenerSubscription?.apply {
            invoke(false)
            keyboardListenerSubscription = null
        }
    }

    override fun onReselected() {
        binding.currencyList.smoothScrollToPosition(0)
    }

    override fun onShown(previousReplaceable: ru.art2000.extensions.fragments.IReplaceableFragment?) {
        if (!model.isFirstUpdateDone) {
            if (model.isUpdateOnFirstTabOpenEnabled()) {
                model.loadData()
            } else {
                updateSnackbar = binding.currencyList.createThemedSnackbar(
                    R.string.message_manually_update,
                    Snackbar.LENGTH_INDEFINITE
                ).apply {
                    setAction(R.string.action_update) { model.loadData() }
                    show()
                }
            }
        }
    }

    override fun onReplaced(nextReplaceable: ru.art2000.extensions.fragments.IReplaceableFragment?) {
        if (updateSnackbar != null) {
            updateSnackbar!!.dismiss()
            updateSnackbar = null
        }
    }

    private fun applyLoadingState(loadingState: LoadingState) {
        setRefreshStatus(loadingState === LoadingState.LOADING_STARTED)
        val messageId = when (loadingState) {
            LoadingState.UNINITIALISED, LoadingState.LOADING_ENDED -> return
            LoadingState.LOADING_STARTED -> R.string.currencies_update_toast
            LoadingState.NETWORK_ERROR -> R.string.currencies_no_internet
            else -> R.string.currencies_update_failed
        }
        Toast.makeText(requireContext(), messageId, Toast.LENGTH_SHORT).show()
    }

    private val titleUpdatedString by lazy { requireContext().getString(R.string.currency_date) }

    private fun setCurrenciesUpdateDate(date: String) {
        binding.toolbar.title = "$titleUpdatedString $date"
    }

    private fun setRefreshStatus(status: Boolean) {
        if (status && binding.refresher.isRefreshing) return
        binding.refresher.isRefreshing = status
    }
}